### PR TITLE
SAML2 new model validation: Signature

### LINKED ISSUE #2950: Implement and test signature validation in SAML2TokenHandler using the new token validation model
As part of the new validation model implementation for SAML2 tokens, we need to validate the signature using ValidationParameters.

Part of https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/issues/2711

### PR DESCRIPTION
**Pull request by iNinja**
# SAML2 new model validation: Signature
- Implemented signature validation for SAML2 tokens using the new validation model
- Added comparison tests to ensure the new validation model behaves like the original.

Note: This re-uses the work introduced for SAML in a previous PR and should not be merged before that one.

Resolves #2950 

Part of #2711

### CODE REVIEW

File: src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateSignature.cs [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+                return ValidationError.NullParameter(
+                    nameof(samlToken),
+                    new StackFrame(true));
+            }
+
+            if (validationParameters is null)
+            {
+                return ValidationError.NullParameter(
+                    nameof(validationParameters),
+                    new StackFrame(true));
```
Comment (brentschmaltz) [inline src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateSignature.cs]: Make sure we use the new model for stack frames here afterwards.

Comment (iNinja) [reply src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateSignature.cs]: I will raise a separate PR to update all the stack frames to use the new model once it is merged.

File: src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateSignature.cs [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+            }
+
+            // Delegate is set by the user, we call it and return the result.
+            if (validationParameters.SignatureValidator is not null)
+                return validationParameters.SignatureValidator(samlToken, validationParameters, null, callContext);
+
+            // If the user wants to accept unsigned tokens, they must implement the delegate
```
Comment (msbw2) [inline src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateSignature.cs]: I think this could be more clear:
```suggestion
            // If the user wants to accept unsigned tokens, they must set validationParameters.SignatureValidator
```

File: src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateSignature.cs [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+            if (samlToken.Assertion.Signature is null)
```
Comment (msbw2) [inline src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateSignature.cs:39]: I think you need to check that `samlToken.Assertion` is not `null`. Otherwise, this results in a null ref.

File: src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateSignature.cs [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+
+            if (validationParameters is null)
+            {
+                return ValidationError.NullParameter(
+                    nameof(validationParameters),
+                    new StackFrame(true));
+            }
+
+            // Delegate is set by the user, we call it and return the result.
+            if (validationParameters.SignatureValidator is not null)
```
Comment (msbw2) [inline src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateSignature.cs:35]: Do you also need to consider `validationParameters.SignatureValidatorUsingConfiguration`?

File: src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateSignature.cs [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+            {
+                resolvedKey = validationParameters.IssuerSigningKeyResolver(
+                    samlToken.Assertion.CanonicalString,
+                    samlToken,
+                    samlToken.Assertion.Signature.KeyInfo?.Id,
+                    validationParameters,
+                    null,
+                    callContext);
+            }
+            else
```
Comment (msbw2) [inline src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateSignature.cs:61]: `validationParameters.IssuerSigningKeyResolverUsingConfiguation`? There are also a few other places where the signing key could be provided and/or validated via `validationParameters`.

File: src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateSignature.cs [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+            }
+
+            if (resolvedKey is null)
+            {
+                if (validationParameters.TryAllIssuerSigningKeys)
+                    keys = validationParameters.IssuerSigningKeys;
+            }
+            else
+            {
+                keys = [resolvedKey];
```
Comment (msbw2) [inline src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateSignature.cs]: Can we just add a block to handle the single key `resolvedKey` instead of allocating a new list with one item?

File: src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateSignature.cs [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+                        validationParameters,
+                        callContext);
+
+                    if (!algorithmValidationResult.IsValid)
+                    {
+                        errors.Add(algorithmValidationResult.UnwrapError());
+                    }
+                    else
+                    {
+                        var validationError = samlToken.Assertion.Signature.Verify(
```
Comment (msbw2) [inline src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateSignature.cs]: Is `Verify` guaranteed to not be `null`?

File: src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateSignature.cs [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+
+                        if (validationError is null)
+                        {
+                            samlToken.SigningKey = key;
+
+                            return key;
+                        }
+                        else
+                        {
+                            errors.Add(validationError.AddStackFrame(new StackFrame()));
```
Comment (msbw2) [inline src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateSignature.cs]: Should `true` be passed to the constructor for `StackFrame` here?

File: src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateSignature.cs [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+                    keys = validationParameters.IssuerSigningKeys;
+            }
+            else
+            {
+                keys = [resolvedKey];
+                keyMatched = true;
+            }
+
+            bool canMatchKey = samlToken.Assertion.Signature.KeyInfo != null;
+            List<ValidationError> errors = new();
```
Comment (msbw2) [inline src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateSignature.cs]: Can you create this only when necessary? If you need to add to it, `(errors ??= new()).Add(...);` would work.

Comment (msbw2) [reply src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateSignature.cs]: Also, what is the expected number of `ValidationError`s? Besides 0, would 1 be the next most common? If so, we can just store the first and then only when we hit 2 create the list.

File: src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateSignature.cs [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+                        }
+                    }
+
+                    keysAttempted.Append(key.ToString());
+                    if (canMatchKey && !keyMatched && key.KeyId is not null && samlToken.Assertion.Signature.KeyInfo is not null)
+                        keyMatched = samlToken.Assertion.Signature.KeyInfo.MatchesKey(key);
+                }
+            }
+
+            if (canMatchKey && keyMatched)
```
Comment (msbw2) [inline src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateSignature.cs:104]: Is this right? Shouldn't it be if `!keyMatched`?
```suggestion
            if (canMatchKey && !keyMatched)
```

File: src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateSignature.cs [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+                        TokenLogMessages.IDX10514,
+                        keysAttempted.ToString(),
+                        samlToken.Assertion.Signature.KeyInfo,
+                        GetErrorStrings(errors),
+                        samlToken),
+                    ValidationFailureType.SignatureValidationFailed,
+                    typeof(SecurityTokenInvalidSignatureException),
+                    new StackFrame(true));
+
+            if (keysAttempted.Length > 0)
```
Comment (msbw2) [inline src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateSignature.cs]: Why is it an error if multiple keys are tried?

File: src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateSignature.cs (hunk 10 above)
Comment (msbw2) [reply src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateSignature.cs:104]: I think I see now. In the valid case this method should have already returned and thus this wouldn't be reached. Therefore, everything at this point is an error.

File: src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateSignature.cs (hunk 11 above)
Comment (msbw2) [reply src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateSignature.cs]: Same here.

File: src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateSignature.cs [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+            return new XmlValidationError(
+                new MessageDetail(TokenLogMessages.IDX10500),
+                ValidationFailureType.SignatureValidationFailed,
+                typeof(SecurityTokenSignatureKeyNotFoundException),
+                new StackFrame(true));
+        }
+
+        private static string GetErrorStrings(List<ValidationError> errors)
+        {
+            StringBuilder sb = new();
```
Comment (msbw2) [inline src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateSignature.cs:186]:
```cs
if (errors.Length == 1)
    return errors[0].ToString();
```

File: src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateSignature.cs [hunk 13]
```diff
@@ -0,0 +1,168 @@
...
+            }
+            else
+            {
+                keys = [resolvedKey];
+                keyMatched = true;
+            }
+
+            bool canMatchKey = samlToken.Assertion.Signature.KeyInfo != null;
+            List<ValidationError> errors = new();
+            StringBuilder keysAttempted = new();
```
Comment (msbw2) [inline src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateSignature.cs]: Is there a `StringBuilder` pool available here?

File: src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateToken.Internal.cs [hunk 14]
```diff
@@ -63,6 +63,13 @@ internal async Task<ValidationResult<ValidatedToken>> ValidateTokenAsync(
                 return validatedIssuerResult.UnwrapError().AddStackFrame(StackFrames.IssuerValidationFailed);
             }
 
+            var signatureValidationResult = ValidateSignature(samlToken, validationParameters, callContext);
+            if (!signatureValidationResult.IsValid)
+            {
+                StackFrames.SignatureValidationFailed ??= new StackFrame(true);
```
Comment (msbw2) [inline src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateToken.Internal.cs:88]: When would this already be set such that we'd end up here?

File: test/Microsoft.IdentityModel.Tokens.Saml.Tests/Saml2SecurityTokenHandlerTests.ValidateTokenAsyncTests.Signature.cs [hunk 15]
```diff
@@ -0,0 +1,211 @@
...
+                    ValidatedToken validatedToken = validationResult.UnwrapResult();
+                    IdentityComparer.AreEqual(validatedToken.SecurityToken, tokenValidationResult.SecurityToken, context);
+                }
+                else
+                {
+                    // Verify the exception provided by both paths match.
+                    var tokenValidationResultException = tokenValidationResult.Exception;
+                    var validationResultException = validationResult.UnwrapError().GetException();
+
+                    if (theoryData.TestId == "Invalid_TokenSignedWithDifferentKey_KeyIdPresent_TryAllKeysFalse")
```
Comment (msbw2) [inline test/Microsoft.IdentityModel.Tokens.Saml.Tests/Saml2SecurityTokenHandlerTests.ValidateTokenAsyncTests.Signature.cs]: Why is this needed? Just left over from debugging?

File: src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateSignature.cs (hunk 4 above)
Comment (iNinja) [reply src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateSignature.cs:35]: `SignatureValidatorUsingConfiguration` only exists in `TokenValidationParameters`. This path uses the new `ValidationParameters` object that has a single `SignatureValidator`

File: test/Microsoft.IdentityModel.Tokens.Saml.Tests/Saml2SecurityTokenHandlerTests.ValidateTokenAsyncTests.Signature.cs (hunk 15 above)
Comment (iNinja) [reply test/Microsoft.IdentityModel.Tokens.Saml.Tests/Saml2SecurityTokenHandlerTests.ValidateTokenAsyncTests.Signature.cs]: Yes. This was for debug purposes.

File: src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateSignature.cs (hunk 3 above)
Comment (iNinja) [reply src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateSignature.cs:39]: Neither `SamlSecurityToken` nor `Saml2SecurityToken` can be instantiated without the assertion. It is always not null.

File: src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateSignature.cs (hunk 5 above)
Comment (iNinja) [reply src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateSignature.cs:61]: This path uses `ValidationParameters` which unified the issuer signing keys into a single list and the delegates into the one `SignatureValidator`.

File: src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateSignature.cs (hunk 7 above)
Comment (iNinja) [reply src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateSignature.cs]: `Verify` is a method on `Signature`, as long as `Signature` exists, `Verify` will always be there.

File: src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateSignature.cs (hunk 8 above)
Comment (iNinja) [reply src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateSignature.cs]: Yes, but we are replacing this in another open PR.

File: src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateToken.Internal.cs (hunk 14 above)
Comment (iNinja) [reply src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateToken.Internal.cs:88]: I am not sure I follow the question. If the signature validation fails, we ensure the stack frame is created, then we return the error including it.
We are moving the caching logic away in a separate PR.

File: src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateSignature.cs (hunk 9 above)
Comment (iNinja) [reply src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateSignature.cs]: Updated to avoid creating the list in a successful scenario.
I'll keep further optimisations for the future, as we have an issue created to refactor the signature validation process further.

File: src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateSignature.cs (hunk 13 above)
Comment (iNinja) [reply src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateSignature.cs]: I don't think so. I changed this to avoid creating the string builder in the successful scenario.

File: src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateSignature.cs (hunk 6 above)
Comment (iNinja) [reply src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateSignature.cs]: Added an optimisation for the case where the key is resolved, separating the degenerate scenario where we iterate over every available key.

File: src/Microsoft.IdentityModel.Tokens.Saml/Saml/SamlSecurityTokenHandler.ValidateSignature.cs [hunk 16]
```diff
@@ -66,57 +65,40 @@ internal static ValidationResult<SecurityKey> ValidateSignature(
...
             }
 
-            bool canMatchKey = samlToken.Assertion.Signature.KeyInfo != null;
-            List<ValidationError> errors = new();
-            StringBuilder keysAttempted = new();
-
             if (keys is not null)
             {
+                // Control reaches here only if the key could not be resolved and TryAllIssuerSigningKeys is set to true.
+                // We try all the keys in the list and return the first valid key. This is the degenerate case.
```
Comment (westin-m) [inline src/Microsoft.IdentityModel.Tokens.Saml/Saml/SamlSecurityTokenHandler.ValidateSignature.cs:86]: > This is the degenerate case.

**Question (non-blocking)** What does "degenerate case" mean in this context? Are there other cases that can be called out in comments to clarify this?

Comment (iNinja) [reply src/Microsoft.IdentityModel.Tokens.Saml/Saml/SamlSecurityTokenHandler.ValidateSignature.cs:86]: I originally borrowed the term from our JWT handler. I read it as a bit of an edge case where we are not focused on obtaining the best performance, as if that were the case, we would have gone with the alternative/trivial case.

In this context, it means that the application was configured to say "if you can't resolve the key, here's potentially a bunch of others you can try, even though that may take time", by enabling `TryAllIssuerKeys` and receiving multiple keys to try.

The expected most common scenario is that the token has a key ID identifying the key it was signed with, we either received from the user or from the configuration a key with that same ID, we match them, verify the signature and either succeed or fail in a single attempt.

**Review by westin-m — APPROVED** [review]

File: src/Microsoft.IdentityModel.Tokens.Saml/Saml/SamlSecurityTokenHandler.ValidateSignature.cs [hunk 17]
```diff
@@ -126,38 +108,88 @@ internal static ValidationResult<SecurityKey> ValidateSignature(
...
+                        callContext);
+
+            if (!algorithmValidationResult.IsValid)
+            {
+                return algorithmValidationResult.UnwrapError().AddCurrentStackFrame();
+            }
+            else
+            {
+                var validationError = samlToken.Assertion.Signature.Verify(
+                key,
```
Comment (brentschmaltz) [inline src/Microsoft.IdentityModel.Tokens.Saml/Saml/SamlSecurityTokenHandler.ValidateSignature.cs]: tab the parameters over one tab.

File: src/Microsoft.IdentityModel.Tokens.Saml/Saml/SamlSecurityTokenHandler.ValidateSignature.cs [hunk 18]
```diff
@@ -66,57 +65,40 @@ internal static ValidationResult<SecurityKey> ValidateSignature(
                 resolvedKey = SamlTokenUtilities.ResolveTokenSigningKey(samlToken.Assertion.Signature.KeyInfo, validationParameters);
             }
 
-            if (resolvedKey is null)
+            bool canMatchKey = samlToken.Assertion.Signature.KeyInfo != null;
+            List<ValidationError>? errors = null;
+            ValidationError? error = null;
+            StringBuilder? keysAttempted = null;
```
Comment (brentschmaltz) [inline src/Microsoft.IdentityModel.Tokens.Saml/Saml/SamlSecurityTokenHandler.ValidateSignature.cs]: Can we move these local variables closer to usage.
For example, keysAttempted many never be used if we return on line 78.

File: src/Microsoft.IdentityModel.Tokens.Saml/Saml/SamlSecurityTokenHandler.ValidateSignature.cs [hunk 19]
```diff
@@ -126,38 +108,88 @@ internal static ValidationResult<SecurityKey> ValidateSignature(
...
                 new MessageDetail(TokenLogMessages.IDX10500),
                 ValidationFailureType.SignatureValidationFailed,
                 typeof(SecurityTokenSignatureKeyNotFoundException),
-                new StackFrame(true));
+                ValidationError.GetCurrentStackFrame());
+        }
+
+        private static ValidationResult<SecurityKey> ValidateSignatureUsingKey(SecurityKey key, SamlSecurityToken samlToken, ValidationParameters validationParameters, CallContext callContext)
+        {
+            ValidationResult<string> algorithmValidationResult = validationParameters.AlgorithmValidator(
```
Comment (SaurabhMSFT) [inline src/Microsoft.IdentityModel.Tokens.Saml/Saml/SamlSecurityTokenHandler.ValidateSignature.cs:143]: `var` here

Comment (iNinja) [reply src/Microsoft.IdentityModel.Tokens.Saml/Saml/SamlSecurityTokenHandler.ValidateSignature.cs:143]: It is clearer to people reading the code if we use the specific return type instead of leaving var and people guessing / inspecting the return type.

File: src/Microsoft.IdentityModel.Tokens.Saml/Saml/SamlSecurityTokenHandler.ValidateSignature.cs [hunk 20]
```diff
@@ -126,38 +108,88 @@ internal static ValidationResult<SecurityKey> ValidateSignature(
...
+                }
+                else
+                {
+                    return validationError.AddCurrentStackFrame();
+                }
+            }
         }
 
-        private static string GetErrorStrings(List<ValidationError> errors)
+        private static string GetErrorStrings(ValidationError? error, List<ValidationError>? errors)
```
Comment (SaurabhMSFT) [inline src/Microsoft.IdentityModel.Tokens.Saml/Saml/SamlSecurityTokenHandler.ValidateSignature.cs]: could you use errorList instead of error? The variable names are too to close to differentiate on casual reading.

Comment (SaurabhMSFT) [inline src/Microsoft.IdentityModel.Tokens.Saml/Saml/SamlSecurityTokenHandler.ValidateSignature.cs]: `GetErrorStrings` -> `GetErrorString` as we are returning a single string.

Comment (brentschmaltz) [inline src/Microsoft.IdentityModel.Tokens.Saml/Saml/SamlSecurityTokenHandler.ValidateSignature.cs]: Several spots in our runtime we build strings, I opened https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/issues/2986 to address.

Comment (iNinja) [reply src/Microsoft.IdentityModel.Tokens.Saml/Saml/SamlSecurityTokenHandler.ValidateSignature.cs]: Thanks. I'm trying to think of a way to handle this. I'll share if I come up with any ideas.

File: src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateSignature.cs [hunk 21]
```diff
@@ -0,0 +1,200 @@
...
+            if (validationParameters.SignatureValidator is not null)
+                return validationParameters.SignatureValidator(samlToken, validationParameters, null, callContext);
+
+            // If the user wants to accept unsigned tokens, they must set validationParameters.SignatureValidator
+            if (samlToken.Assertion.Signature is null)
+                return new XmlValidationError(
+                    new MessageDetail(
+                        TokenLogMessages.IDX10504,
+                        samlToken.Assertion.CanonicalString),
+                    ValidationFailureType.SignatureValidationFailed,
```
Comment (brentschmaltz) [inline src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateSignature.cs]: Could we return a new ValidationFailureType?
MissingSignatureFailureType.

File: src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateSignature.cs [hunk 22]
```diff
@@ -0,0 +1,200 @@
...
+                    typeof(SecurityTokenValidationException),
+                    ValidationError.GetCurrentStackFrame());
+
+            IList<SecurityKey>? keys = null;
```
Comment (brentschmaltz) [inline src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateSignature.cs]: As in SamlSecurityTokenHandler, it is easier to review if local variables are declared just before they are used.

File: src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateSignature.cs [hunk 23]
```diff
@@ -0,0 +1,200 @@
...
+
+                error = result.UnwrapError();
+            }
+            else
+            {
+                if (validationParameters.TryAllIssuerSigningKeys)
+                    keys = validationParameters.IssuerSigningKeys;
+            }
+
+            if (keys is not null)
```
Comment (brentschmaltz) [inline src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateSignature.cs]: We may not need the 'keys' local variable.
Could we write
```cs
if (validationParameters.TryAllIssuerSigningKeys && resolvedKey == null)
    for (int i=0; i < validationParameters.IssuerSigningKeys.Count; i++)
    {

    }
```

May be able to remove the else clause above.

File: src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateSignature.cs [hunk 24]
```diff
@@ -0,0 +1,200 @@
...
+            {
+                // Control reaches here only if the key could not be resolved and TryAllIssuerSigningKeys is set to true.
+                // We try all the keys in the list and return the first valid key. This is the degenerate case.
+                for (int i = 0; i < keys.Count; i++)
+                {
+                    SecurityKey key = keys[i];
```
Comment (brentschmaltz) [inline src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateSignature.cs]: A key could be null, we should check.
Could we write
```cs
if (keys[i] != null)
{
    var result = ValidateSignatureUsingKey(keys[i], samlToken, validationParameters, callContext);
    ...
}
```
Also skipping the local variable 'key'.

File: src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateSignature.cs [hunk 25]
```diff
@@ -0,0 +1,200 @@
...
+                        callContext);
+
+            if (!algorithmValidationResult.IsValid)
+            {
+                return algorithmValidationResult.UnwrapError().AddCurrentStackFrame();
+            }
+            else
+            {
+                var validationError = samlToken.Assertion.Signature.Verify(
+                key,
```
Comment (brentschmaltz) [inline src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateSignature.cs]: tab over parameters.

**Review by brentschmaltz — CHANGES_REQUESTED** [review]
Some small requests, looks good.

File: src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateSignature.cs [hunk 26]
```diff
@@ -0,0 +1,200 @@
...
+                    return key;
+                }
+                else
+                {
+                    return validationError.AddCurrentStackFrame();
+                }
+            }
+        }
+
+        private static string GetErrorStrings(ValidationError? error, List<ValidationError>? errors)
```
Comment (SaurabhMSFT) [inline src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateSignature.cs:173]: This seems like a copy of method in the previous file. Is there any reason we are duplicating the code.

File: test/Microsoft.IdentityModel.Tokens.Saml.Tests/Saml2SecurityTokenHandlerTests.ValidateTokenAsyncTests.Signature.cs [hunk 27]
```diff
@@ -0,0 +1,207 @@
...
+
+                    theoryData.ExpectedException.ProcessException(tokenValidationResultException, context);
+                    theoryData.ExpectedExceptionValidationParameters!.ProcessException(validationResultException, context);
+                }
+
+                TestUtilities.AssertFailIfErrors(context);
+            }
+        }
+
+        public static TheoryData<ValidateTokenAsyncSignatureTheoryData> ValidateTokenAsync_Signature_TestCases
```
Comment (SaurabhMSFT) [inline test/Microsoft.IdentityModel.Tokens.Saml.Tests/Saml2SecurityTokenHandlerTests.ValidateTokenAsyncTests.Signature.cs:65]: this should be private static method instead of a property

**Review by SaurabhMSFT — DISMISSED** [review]
.

File: src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateSignature.cs (hunk 26 above)
Comment (brentschmaltz) [reply src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateSignature.cs:173]: Agreed this could be common, nothing token specific about it.

Comment (iNinja) [reply src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateSignature.cs:173]: Yes, because the SAML2 code is almost a copy of the SAML one, but we are not in a position to refactor them into anything different at this time.

File: test/Microsoft.IdentityModel.Tokens.Saml.Tests/Saml2SecurityTokenHandlerTests.ValidateTokenAsyncTests.Signature.cs (hunk 27 above)
Comment (iNinja) [reply test/Microsoft.IdentityModel.Tokens.Saml.Tests/Saml2SecurityTokenHandlerTests.ValidateTokenAsyncTests.Signature.cs:65]: As I replied in a different PR where you suggested the same about making the TheoryData private, XUnit requires this to be public.

Can you elaborate on on the reason why this should be a method instead of a property? We have all of our tests written using this as a property.

File: src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateSignature.cs (hunk 26 above)
Comment (iNinja) [reply src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateSignature.cs:173]: @brentschmaltz unless we do some refactor to share code between SAML and SAML2, I don't see the point in introducing code sharing for one helper method.

File: src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateSignature.cs (hunk 21 above)
Comment (iNinja) [reply src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateSignature.cs]: Added a type, but following the naming you used in the extensibility `IssuerValidationDelegateThrew`, I think that is a good naming convention making this one `TokenIsNotSigned`. I don't think we need `FailureType` in the name.

File: src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateSignature.cs (hunk 24 above)
Comment (iNinja) [reply src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateSignature.cs]: We use this variable in 4 places in this method. I think I'd rather keep it.

Adding the null check.

File: src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateSignature.cs (hunk 23 above)
Comment (iNinja) [reply src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.ValidateSignature.cs]: Removed the `keys` local variable and simplified the code a bit. We do need to check for `keyMatched` to avoid re-using the key if it was resolved.

**Review by brentschmaltz — APPROVED** [review]
